### PR TITLE
Fixing segfault if filtered_cloud_topic is not set in PointCloudOctomapUpdater

### PR DESCRIPTION
**Pull request by jasedit**
Iterators were created into a PointCloud2 message which was only created if filtered_cloud_topic was set. Furthermore, the incoming data was never used during the update - the octomap update iterated over the output PointCloud2, as opposed to the input PointCloud 2. On my setup (Trusty/Indigo), PointCloudOctomapUpdater would crash immediately unless I defined filtered_cloud_topic, and I would never see any output data. I believe this pull request fixes this issue.

### CODE REVIEW

**Comment by jasedit** [comment]
Odd - the build error is that perception/point_containment_filter/src/shape_mask.cpp could not find sensor_msgs/point_cloud2_iterator.h, which are not files I've touched in any way. It looks like CMakeLists.txt and packages.xml did not declare any dependency on sensor_msgs, which I've added. I think that should fix the Travis build error.

**Comment by mikeferguson** [comment]
@jasedit Disregard build errors -- Travis doesn't support Trusty, so this is still being built against precise/hydro, and the iterator does not exist in hydro.

**Comment by mikeferguson** [comment]
Looping in @vrabaud since he created the point_cloud2_iterator.

**Comment by davetcoleman** [comment]
I've just run into this issue, I'll test your PR

Also, I've just disabled Travis for all MoveIt! repos per the mailing list email I sent out.

**Comment by jasedit** [comment]
Thanks! I've verified that it no longer crashes on my system. However, I can't validate any of the data coming out of the updater due to #503, which I haven't made any headway on solving.

**Comment by davetcoleman** [comment]
I'm still getting a segfault even with your branch when I start move_group with a point cloud being published:

> 0x00007fffd1abd503 in occupancy_map_monitor::PointCloudOctomapUpdater::cloudMsgCallback (this=0xa25f40, cloud_msg=...) at /home/dave/ros/ws_moveit/src/moveit_ros/perception/pointcloud_octomap_updater/src/pointcloud_octomap_updater.cpp:249

I'm currently debugging but any hints are welcomed

**Comment by davetcoleman** [comment]
Also, it is not clear to me why you made the changes you made in this PR. Perhaps adding inline comments to your changes would help me and others in the future?

**Comment by jasedit** [comment]
Line 249 would imply that cloud_msg is an empty/invalid message. There doesn't appear to be a check for this, and may be a reason why I haven't seen it yet.

**Comment by jasedit** [comment]
I can add comments on the iterators momentarily.

**Comment by jasedit** [comment]
Comments added - if there is additional info necessary, let me know and I'll be happy to flesh it out.

**Comment by davetcoleman** [comment]
Thanks for the comments, I read a lot into the code just now. I believe your fixes are good but I think there are some more issues, in fact. See https://github.com/ros-planning/moveit_ros/issues/505.

In your testing are you actually filtering a point cloud, or do you have that feature disabled?

**Comment by jasedit** [comment]
I am filtering a point cloud using this block of code. I don't think you can cause the segfault without filtering the points.

File: perception/pointcloud_octomap_updater/src/pointcloud_octomap_updater.cpp [hunk 1]
```diff
@@ -224,46 +233,45 @@ void PointCloudOctomapUpdater::cloudMsgCallback(const sensor_msgs::PointCloud2::
...
-      sensor_msgs::PointCloud2Iterator<float> iter_z(*filtered_cloud, "z");
-      iter_x += row*cloud_msg->row_step;
-      iter_y += row*cloud_msg->row_step;
-      iter_z += row*cloud_msg->row_step;
-
-      for (unsigned int col = 0; col < cloud_msg->width; col += point_subsample_, iter_x += point_subsample_,
-        iter_y += point_subsample_, iter_z += point_subsample_)
+      sensor_msgs::PointCloud2ConstIterator<float> pt_iter(*cloud_msg, "x");
+      //set iterator to point at start of the current row
+      pt_iter += row;
```
Comment (vrabaud) [inline perception/pointcloud_octomap_updater/src/pointcloud_octomap_updater.cpp]: that is wrong, you should have 

```
t_iter += row*cloud_msg->width;
```

Comment (jasedit) [inline perception/pointcloud_octomap_updater/src/pointcloud_octomap_updater.cpp]: Or `pt_iter += row_c`, no?

Comment (vrabaud) [inline perception/pointcloud_octomap_updater/src/pointcloud_octomap_updater.cpp]: that works too.

**Comment by jasedit** [comment]
Commit is fixed to correctly address #505, I believe.

**Comment by mikeferguson** [comment]
Ok, here's my thoughts:
- +1 to all changes in pointcloud_octomap_updater.cpp
- +1 to adding depend on sensor_msgs
- -0 to adding depends in cmake on sensor_msgs_generate_messages_cpp, there are a lot of other things that aren't declared, and most people aren't building sensor_msgs locally

I would recommend squashing this down to a single commit also.

**Comment by jasedit** [comment]
Commits have been squashed, sensor_msgs_generate_messages_cpp has been removed.

File: perception/point_containment_filter/CMakeLists.txt [hunk 2]
```diff
@@ -4,6 +4,7 @@ add_library(${MOVEIT_LIB_NAME} src/shape_mask.cpp)
 set_target_properties(${MOVEIT_LIB_NAME} PROPERTIES COMPILE_FLAGS "${CMAKE_CXX_FLAGS} ${OpenMP_CXX_FLAGS}")
 set_target_properties(${MOVEIT_LIB_NAME} PROPERTIES LINK_FLAGS "${OpenMP_CXX_FLAGS}")
 target_link_libraries(${MOVEIT_LIB_NAME} ${catkin_LIBRARIES} ${Boost_LIBRARIES})
+add_dependencies(${MOVEIT_LIB_NAME} sensor_msgs_generate_messages_cpp)
```
Comment (davetcoleman) [inline perception/point_containment_filter/CMakeLists.txt]: You said you took this out, right?

File: perception/pointcloud_octomap_updater/src/pointcloud_octomap_updater.cpp [hunk 3]
```diff
@@ -224,46 +233,45 @@ void PointCloudOctomapUpdater::cloudMsgCallback(const sensor_msgs::PointCloud2::
   {
     /* do ray tracing to find which cells this point cloud indicates should be free, and which it indicates
      * should be occupied */
-    for (unsigned int row = 0; row < cloud_msg->height; row += point_subsample_)
+    for(unsigned int row = 0; row < cloud_msg->height; row += point_subsample_)
```
Comment (davetcoleman) [inline perception/pointcloud_octomap_updater/src/pointcloud_octomap_updater.cpp]: This is super minor, but a digression from this file's standard

**Comment by davetcoleman** [comment]
It works - thanks so much! I'll merge as soon as my two comments are addressed.

**Comment by jasedit** [comment]
Addressed!

**Comment by jasedit** [comment]
Thanks for helping get to the bottom of this!

**Pull request merged**